utils/go-client/metric.go

```diff
@@ -0,0 +1,217 @@
+package go_client
+
+import (
+	"io/ioutil"
+	"net/http"
+	"path"
+	"strings"
+
+	"github.com/prometheus/common/log"
+	"k8s.io/client-go/rest"
+
+	"github.com/argoproj/argo-rollouts/utils/unstructured"
+)
+
+type K8sRequestVerb string
+
+const (
+	List    K8sRequestVerb = "List"
+	Get     K8sRequestVerb = "Get"
+	Create  K8sRequestVerb = "Create"
+	Update  K8sRequestVerb = "Update"
+	Patch   K8sRequestVerb = "Patch"
+	Delete  K8sRequestVerb = "Delete"
+	Unknown K8sRequestVerb = "Unknown"
+)
+
+type ResourceInfo struct {
+	Kind       string
+	Namespace  string
+	Name       string
+	Verb       K8sRequestVerb
+	StatusCode int
+}
+
+func (ri ResourceInfo) HasAllFields() bool {
+	return ri.Kind != "" && ri.Namespace != "" && ri.Name != "" && ri.Verb != "" && ri.StatusCode != 0
+}
+
+type metricsRoundTripper struct {
+	roundTripper    http.RoundTripper
+	inc             func(ResourceInfo) error
+	commonResources map[string]bool
+}
+
+func (m metricsRoundTripper) resolveK8sRequestVerb(r *http.Request) K8sRequestVerb {
+	if r.Method == "POST" {
+		return Create
+	}
+	if r.Method == "DELETE" {
+		return Delete
+	}
+	if r.Method == "PATCH" {
+		return Patch
+	}
+	if r.Method == "PUT" {
+		return Update
+	}
+	if r.Method == "GET" {
+		resource := path.Base(r.URL.Path)
+		if _, ok := m.commonResources[resource]; ok {
+			return List
```

Comment:
Why are we assuming that if is only common resources is it a list request? We can't infer based on url fragments?